tensorflow/c/experimental/gradients/model_factory_helper.h

```diff
@@ -0,0 +1,98 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+#ifndef TENSORFLOW_C_EXPERIMENTAL_GRADIENTS_MODEL_FACTORY_MACRO_H_
+#define TENSORFLOW_C_EXPERIMENTAL_GRADIENTS_MODEL_FACTORY_MACRO_H_
+
+// This macro will expand to a function that defines a `Model`. This `Model` is
+// then used for testing by `nn_grad_test` and `math_grad_test`. `ops_call` is a
+// statement that calls to a `ops::` and should be wrapped around by `{}`.
+// `ops_call` has access to `inputs`. The output parameter of the ops should
+// always be `absl::MakeSpan(temp_outputs)`. This macro supports most one-ops
+// model.
+// TODO(vnvo2409): Extend support for more complex model.
+#define TF_MODEL_FACTORY(name, num_inputs, num_outputs, ops_call)      \
+  Status name(AbstractContext* ctx,                                    \
+              absl::Span<AbstractTensorHandle* const> inputs,          \
+              absl::Span<AbstractTensorHandle*> outputs,               \
+              const GradientRegistry& registry) {                      \
+    auto tape = new Tape(/*persistent=*/false);                        \
```

Comment:
Since we do not wish to compute gradients here, this function is effectively just calling `ops_call`.

Reply:
Agree. However, it saves us a lot of copy paste ( like new Tape, Watch, ... ) so I think this macro might be useful. In addition, a model without computing gradients is needed for `gradient_checker`.

Reply:
Since we never call `tape->ComputeGradient` here, `tape->Watch`, `TapeContext` etc are no-ops. So this macro is effectively:

```
#define TF_MODEL_FACTORY(name, num_inputs, num_outputs, ops_call)      \
  Status name(AbstractContext* ctx,                                    \
              absl::Span<AbstractTensorHandle* const> inputs,          \
              absl::Span<AbstractTensorHandle*> outputs,               \
              const GradientRegistry& registry) {                      \
    ops_call;                                                          \
    return Status::OK();                                               \
}
```

Reply:
Since the old models use `tape->ComputeGradient`, `tape->Watch`, etc, I just copy without thinking too much. If it is this much simple, we don't need a macro.